---
layout: single
published: true
title:  "[JAVA] 자동 리소스 닫기(try-with-resource)"
categories: Java
date: 2024-02-07 18:56:53
toc: true
toc_sticky: true
tag:   
  - Java
  - 자바
  - 자동 리소스 닫기(try-with-resource)
---

자바 7에서 새롭게 추가된 try-with-resource를 사용하면 **예외 발생 여부와 상관없이** 사용했던 리소스 객체(입출력, 스트림, 서버소켓, …)의 close() 메소드를 호출해서 안전하게 리소스를 닫아줍니다. 여기서 리소스는 다양한 의미가 될 수 있는데, 데이터를 읽고 쓰는 객체 라고 이해를 하시면 됩니다. 예를들어 파일의 데이터를 읽는 `BufferedReader` 객체와 파일 객체인 `FileReader`는 리소스 객체라고 보면 됩니다. 

다음 코드는 자바6 이전까지 사용했던, 리소스 객체를 안전하게 닫기 위한 코드입니다.

```java
import java.io.BufferedReader;
import java.io.FileNotFoundException;
import java.io.FileReader;
import java.io.IOException;

public class TryWithResourceExample {

		public static void main(String[] args) {
					BufferedReader br = null;
	        try {
	            br = new BufferedReader(new FileReader("file.txt"));
	            br.readLine();
	            br.close();
	        } catch (FileNotFoundException e) {
	            // FileNotFoundException 발생 예외 처리
	        } catch (IOException e) {
	            // IOException 발생 예외 처리
	        } finally {
	            if (br != null) {
	                try {
	                    br.close();
	                } catch (IOException e) {
	                    // 예외처리
	                }
	            }
	        }
		}
}
```

finally 블록에서 다시 try-catch를 열고 최종적으로 BufferedReader 리소스 close를 해주죠. close 하는 과정에서 IOException이 발생할 가능성이 있기 때문에 finally 블록 안에서도 try-catch를 사용해야합니다. 많이 복잡해보입니다.
그럼 이부분을 자바7에서 추가된 try-with-resource구문을 사용해서 리소스 close를 해주겠습니다.

```java
try (BufferedReader reader = new BufferedReader(new FileReader("file.txt"))) {
	...

} catch (IOException e) {
	// 예외 처리

}
```

정말 간단해지죠? 어디를 봐도 close()를 명시적으로 호출한 곳이 없습니다.

try 블록이 정상적으로 실행을 완료했거나 혹은 중간에 예외가 발생되면 자동으로 BufferedReader의 close() 메소드가 호출됩니다. try {} 에서 예외가 발생하면 우선 close()로 리소스를 닫고, catch 블록을 실행합니다. 만약 복수개의 리소스를 사용해야한다면 다음과 같이 작성할 수 있습니다.

```java
try (BufferedReader reader = new BufferedReader(new FileReader("file.txt"));
     BufferedWriter writer = new BufferedWriter(new FileWriter("file2.txt"))) {

} catch (IOException e) {
    // 예외 처리

}
```

```java
try (FileInputStream inputStream = new FileInputStream("file.txt");
     FileOutputStream outputStream = new FileOutputStream("file2.txt")) {
	...

} catch (IOException e) {
	// 예외 처리

}
```

### try-with-resource 사용 조건

이렇게 간단한 자동 리소스 닫기 기능을 사용하는데는 필요한 조건이 있습니다. 리소스 객체는 **java.lang.AutoCloseable** 인터페이스를 구현하고 있어야 합니다. AutoCloseable에는 close() 메소드가 정의되어 있는데 try-with-resources는 바로 이 close() 메소드를 자동 호출합니다. 

자바 API 도큐먼트에서 **AutoCloseable** 인터페이스를 찾아서 “All Known Implementing Classes:”를 보면 try-with-resources와 함께 사용할 수 있는 리소스가 어떤것들이 있는지 알 수 있습니다. 

[Java Platform SE 8](https://docs.oracle.com/javase/8/docs/api/index.html?overview-summary.html)

![image](https://github.com/BaxDailyGit/BaxDailyGit/assets/99312529/740929e8-9a5a-400e-a33b-629a46560c5f)

(참고) Closeable, AutoCloseable 인터페이스

`AutoCloseable` 인터페이스의 close() 메소드를 직접 구현하는 `FileInputStream` 클래스를 작성해보겠습니다. 그리고 `TryWithResourceExample` 코드를 보고 출력결과를 한번 생각해보세요.

```java
public class FileInputStream implements AutoCloseable {
	private String file;

	public FileInputStream(String file) {
		this.file = file;
	}

	public void read() {
		System.out.println(file + " 을 읽습니다");
	}

	@Override
	public void close() throws Exception {
		System.out.println(file + "을 닫습니다");
	}
}
```

```java
public class TryWithResourceExample {
	
	public static void main(String[] args) {
		try (FileInputStream fileInputStream = new FileInputStream("file.txt")){
			fileInputStream.read();
			throw new Exception();		// 강제적으로 예외 발생시킴
		} catch (Exception e) {
			System.out.println("예외 처리 코드가 실행되었습니다.");
		}

	}
}
```

.

.

.

출력결과

```java
file.txt 을 읽습니다
file.txt을 닫습니다
예외 처리 코드가 실행되었습니다.
```